### PR TITLE
[Enhancement] date_diff function support year/month/quarter/week

### PR DESCRIPTION
**Pull request by xuzifu666**
Fixes #issue

## What type of PR is this:
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by xuzifu666** [comment]
@wangsimo0 had add week/year/quarter/month arguments to enhance date_diff function

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: test/sql/test_time_fn/T/test_time_fn [hunk 1]
```diff
@@ -204,6 +204,10 @@ select date_diff("2010-11-30 23:59:59", "2010-11-30 20:58:59", "minute");
 select date_diff("2010-11-30 23:59:59", "2010-11-30 20:59:59", "second");
 select date_diff('2010-11-30 23:59:59.000001', '2010-11-30 20:58:59.000001', 'hour');
 select date_diff('2010-11-30 23:59:59.000001', '2010-11-30 20:58:59.000001', 'day');
+select date_diff('2010-12-22 00:59:59.000001', '2010-11-30 20:58:59.000001', 'week');
+select date_diff('2011-12-22 00:59:59.000001', '2010-11-30 20:58:59.000001', 'month');
+select date_diff('2012-12-22 00:59:59.000001', '2010-11-30 20:58:59.000001', 'year');
+select date_diff('2011-12-22 00:59:59.000001', '2010-11-30 20:58:59.000001', 'quarter');
```
Comment (silverbullet233) [inline test/sql/test_time_fn/T/test_time_fn:210]: need more stronger cases

File: be/src/exprs/time_functions.cpp [hunk 2]
```diff
@@ -2597,6 +2597,14 @@ StatusOr<ColumnPtr> TimeFunctions::datediff(FunctionContext* context, const Colu
             result.append(l.diff_microsecond(r) / USECS_PER_MILLIS);
         } else if (type_str == "day") {
             result.append(l.diff_microsecond(r) / USECS_PER_DAY);
+        } else if (type_str == "week") {
+            result.append(l.diff_microsecond(r) / USECS_PER_WEEK);
+        } else if (type_str == "quarter") {
+            result.append(months_diffImpl::apply<TimestampValue, TimestampValue, int>(l, r) / 3);
```
Comment (silverbullet233) [inline be/src/exprs/time_functions.cpp:2603]: the current implementation of months_diff didn't consider microseconds, need to add more test caes to confim that if we need do some change to meet our requirements

**Review by silverbullet233 — APPROVED** [review]

**Review by imay — APPROVED** [review]

**Pull request merged**